src/rust/ide/src/transport/web.rs

```diff
@@ -98,60 +103,198 @@ impl State {
+    /// Special callback on "close" event. Must be invoked after `on_close`.
+    pub on_close_internal : Slot<event::Close>,
+    /// When enabled, the WS will try to automatically reconnect whenever connection is lost.
+    pub reconnect         : bool,
+}
+
+impl Model {
     /// Wraps given WebSocket object.
-    pub fn new
-    (ws:web_sys::WebSocket, parent:impl AnyLogger, name:impl AsRef<str>) -> WebSocket {
+    pub fn new(ws:web_sys::WebSocket, logger:Logger) -> Model {
         ws.set_binary_type(BinaryType::Arraybuffer);
-        WebSocket {
+        Model {
+            on_close          : Slot::new(&ws, &logger),
+            on_message        : Slot::new(&ws, &logger),
+            on_open           : Slot::new(&ws, &logger),
+            on_error          : Slot::new(&ws, &logger),
+            on_close_internal : Slot::new(&ws, &logger),
+            reconnect         : true,
+            logger,
             ws,
-            logger     : Logger::sub(parent,name),
-            on_message : default(),
-            on_close   : default(),
-            on_open    : default(),
-            on_error   : default(),
         }
     }
 
+    /// Close the socket.
+    pub fn close(&mut self, reason:&str) -> Result<(),JsValue> {
+        self.ws.close_with_code_and_reason(1000,reason)?;
+        self.clear_callbacks();
+        Ok(())
+    }
+
+    /// Clear all the available callbacks.
+    pub fn clear_callbacks(&mut self) {
+        let Self{on_close,on_error,on_message,on_open,on_close_internal,..} = self;
+        // We don't care if removing actually removed anything.
+        // If callbacks were not set, then they are clear from the start.
+        on_close.clear_callback();
+        on_error.clear_callback();
+        on_message.clear_callback();
+        on_open.clear_callback();
+        on_close_internal.clear_callback()
+    }
+
+    /// Establish a new WS connection, using the same URL as the previous one.
+    /// All callbacks will be transferred to the new connection.
+    pub fn reconnect(&mut self) -> Result<(),JsValue> {
+       if !self.reconnect {
+           return Err(js_sys::Error::new("Reconnecting has been disabled").into());
+       }
+
+        let url = self.ws.url();
+        warning!(self.logger, "Reconnecting WS to {url}");
+
+        let new_ws = web_sys::WebSocket::new(&url)?;
+
+        self.on_close.         set_target(&new_ws);
+        self.on_error.         set_target(&new_ws);
+        self.on_message.       set_target(&new_ws);
+        self.on_open.          set_target(&new_ws);
+        self.on_close_internal.set_target(&new_ws);
+        self.ws = new_ws;
+
+        Ok(())
+    }
+}
+
+impl Drop for Model {
+    fn drop(&mut self) {
+        info!(self.logger, "Dropping WS model");
+        if let Err(e) = self.close("Rust Value has been dropped.") {
+            error!(self.logger,"Error when closing socket due to being dropped: {js_to_string(&e)}")
```

Comment:
missing dot

Reply:
The dot is expected from the pretty-printed JS error.